posthog/temporal/batch_exports/transformer.py

```diff
@@ -0,0 +1,230 @@
+                # "\ud83d"
+                logger.exception("Failed to encode with orjson: %s", d)
+                cleaned_content = replace_broken_unicode(d)
+                data_gen = self._write(orjson.dumps(cleaned_content, default=str) + b"\n")
+        yield from data_gen
+
+    def write_batch(self, record_batch: pa.RecordBatch) -> typing.Generator[bytes, None, None]:
+        """Write records to a temporary file as JSONL."""
+        for record_dict in record_batch.to_pylist():
+            if not record_dict:
+                continue
+
+            yield from self._write_dict(record_dict)
+
+    def _write(self, content: bytes | str):
+        """Write bytes to underlying file keeping track of how many bytes were written."""
+        compressed_content = self.compress(content)
+        yield compressed_content
+
+
+class ParquetStreamTransformer(StreamTransformer):
+    def __init__(
+        self,
+        schema: pa.Schema,
+        compression: str | None = None,
+        compression_level: int | None = None,
+        include_inserted_at: bool = False,
+    ):
+        super().__init__(compression=compression, include_inserted_at=include_inserted_at)
+        self.schema = schema
+        self.compression_level = compression_level
+
+        # For Parquet, we need to handle schema and batching
+        self._parquet_writer: pq.ParquetWriter | None = None
+        self._parquet_buffer = BytesIO()
+
+    @property
+    def parquet_writer(self) -> pq.ParquetWriter:
+        if self._parquet_writer is None:
+            self._parquet_writer = pq.ParquetWriter(
+                self._parquet_buffer,
+                schema=self.schema,
+                compression="none" if self.compression is None else self.compression,  # type: ignore
+                compression_level=self.compression_level,
+            )
+        assert self._parquet_writer is not None
+        return self._parquet_writer
+
+    def finalize(self) -> typing.Generator[bytes, None, None]:
+        """Ensure underlying Parquet writer is closed before flushing and closing temporary file."""
+        yield from super().finalize()
+
+        if self._parquet_writer is not None:
+            self._parquet_writer.close()
+            self._parquet_writer = None
+
+            # Get final data without copying
+            final_data = self._parquet_buffer.getvalue()
+            if final_data:
+                yield final_data
+
+            # Cleanup
+            self._parquet_buffer.close()
+            self._parquet_buffer = BytesIO()
+
+    def write_batch(self, record_batch: pa.RecordBatch) -> typing.Generator[bytes, None, None]:
+        """Write records to a temporary file as Parquet."""
+
+        self.parquet_writer.write_batch(record_batch.select(self.parquet_writer.schema.names))
+        # Get current buffer content without copying
+        data = self._parquet_buffer.getvalue()
```

Comment:
This doesn't make sense: The comment says that this is getting "current buffer content without copying", but in the next two lines we cumplete delete everything from the buffer.

How, exactly, are we getting the data from the buffer without copying it? Data must live somewhere, but if it is still in the buffer then we delete it just below, and if it's not in the buffer, then we copied it somewhere?

Reply:
yes, you're right, it does actually copy the data. I've removed this misleading comment. I'm not sure if there is a better way to handle this